File: skip-the-dishes-command-resource/src/main/java/br/com/skip/the/dishes/command/resource/product/ProductApi.java
package br.com.skip.the.dishes.command.resource.product;

import org.springframework.web.bind.annotation.*;

import javax.validation.Valid;
import javax.validation.constraints.NotNull;
import javax.validation.constraints.Size;

import static br.com.skip.the.dishes.command.resource.commons.Paths.PRODUCT_PATH;
import static br.com.skip.the.dishes.domain.utils.DomainConstants.PRODUCT_ID_SIZE;
import static org.springframework.http.HttpStatus.CREATED;
import static org.springframework.http.HttpStatus.OK;

@RequestMapping(path = PRODUCT_PATH)
public interface ProductApi {

    @ResponseStatus(CREATED)
    @PostMapping
    CreateResponse create(@RequestBody @Valid CreateRequest createRequest);

    @ResponseStatus(OK)
    @PatchMapping(path = "/{id}/detail")
    void updateDetail(@PathVariable("id") @Valid @NotNull @Size(max = PRODUCT_ID_SIZE) String id,
                      @RequestBody @Valid DetailRequest detailRequest);

    @ResponseStatus(OK)
    @PatchMapping(path = "/{id}/price")
    void updatePrice(@PathVariable("id") @Valid @NotNull @Size(max = PRODUCT_ID_SIZE) String id,
                     @RequestBody @Valid PriceRequest priceRequest);

}

File: skip-the-dishes-command-resource/src/main/kotlin/br/com/skip/the/dishes/command/resource/product/ProductApi.kt
package br.com.skip.the.dishes.command.resource.product

import br.com.skip.the.dishes.command.resource.commons.Paths.PRODUCT_PATH
import br.com.skip.the.dishes.domain.utils.DomainConstants.PRODUCT_ID_SIZE
import org.springframework.http.HttpStatus.CREATED
import org.springframework.http.HttpStatus.OK
import org.springframework.web.bind.annotation.*
import javax.validation.Valid
import javax.validation.constraints.NotNull
import javax.validation.constraints.Size

@RequestMapping(path = [PRODUCT_PATH])
interface ProductApi {

    @ResponseStatus(CREATED)
    @PostMapping
    fun create(@RequestBody @Valid createProductRequest: CreateProductRequest): CreateProductResponse

    @ResponseStatus(OK)
    @PatchMapping(path = ["/{id}/detail"])
    fun updateDetail(@PathVariable("id") @Valid @NotNull @Size(max = PRODUCT_ID_SIZE) id: String,
                     @RequestBody @Valid detailRequest: DetailRequest)

    @ResponseStatus(OK)
    @PatchMapping(path = ["/{id}/price"])
    fun updatePrice(@PathVariable("id") @Valid @NotNull @Size(max = PRODUCT_ID_SIZE) id: String,
                    @RequestBody @Valid priceRequest: PriceRequest)

}
